os/fs/smartfs/smartfs_utils.c

```diff
@@ -1716,139 +1962,255 @@ int smartfs_countdirentries(struct smartfs_mountpt_s *fs, struct smartfs_entry_s
+	ret = FS_IOCTL(fs, BIOC_READSECT, (unsigned long)&readwrite);
+	if (ret < 0) {
+		fdbg("Unable to read sector %d while extending file, error = %d\n", readwrite.logsector, ret);
+		result = ret;
+		goto errout_with_chain;
+	}
+	chainheader = (struct smartfs_chain_header_s *)fs->fs_rwbuffer;
+	*(uint16_t *)chainheader->nextsector = save_first;
+	/* Append '0's to the remaining part of the file's last sector */
+	if (new_chain) {
+		*(uint16_t *)chainheader->used = data_len;
+		memset(&fs->fs_rwbuffer[sf->curroffset], 0, fs->fs_llformat.availbytes - sf->curroffset);
+	} else {
+		*(uint16_t *)chainheader->used += remaining;
+		memset(&fs->fs_rwbuffer[sf->curroffset], 0, (remaining + offset - sf->curroffset));
+	}
 
-		memset(sf->buffer, CONFIG_SMARTFS_ERASEDSTATE, fs->fs_llformat.availbytes);
-		header = (struct smartfs_chain_header_s *)sf->buffer;
-		header->type = SMARTFS_SECTOR_TYPE_FILE;
-		sf->bflags = SMARTFS_BFLAG_DIRTY;
-		entry->datalen = 0;
+	ret = FS_IOCTL(fs, BIOC_WRITESECT, (unsigned long)&readwrite);
+	if (ret < 0) {
+		fdbg("Unable to write current last file sector %d while extending file, error = %d\n", readwrite.logsector, ret);
+		result = ret;
+		goto errout_with_chain;
 	}
 #endif
 
-	return OK;
-}
-
-/****************************************************************************
- * Name: smartfs_shrinkfile
- *
- * Description:
- *   Shrink the size of an existing file to the specified length
- *
- ****************************************************************************/
-
-int smartfs_shrinkfile(FAR struct smartfs_mountpt_s *fs, FAR struct smartfs_ofile_s *sf, off_t length)
-{
-	/* TODO This will be added later */
-	return OK;
-}
-
-/****************************************************************************
- * Name: smartfs_extendfile
- *
- * Description:
- *   Zero-extend the length of a regular file to 'length'.
- *
- ****************************************************************************/
+errout_with_chain:
+	if (result != OK) {
+		/* A chain of new sectors was created to extend the file but not completed when an error was encountered. Free the chain */
+		chainheader = (struct smartfs_chain_header_s *)fs->fs_rwbuffer;
+		smartfs_setbuffer(&readwrite, save_first, 0, sizeof(struct smartfs_chain_header_s), (FAR uint8_t *)fs->fs_rwbuffer);
+		while (save_first != 0xFFFF) {
+			ret = FS_IOCTL(fs, BIOC_READSECT, (unsigned long)&readwrite);
+			if (ret < 0) {
+				fdbg("Unable to read sector %d to free it, error = %d\n", readwrite.logsector, ret);
+				return ret;
+			}
+			nextsector = SMARTFS_NEXTSECTOR(chainheader);
+			ret = FS_IOCTL(fs, BIOC_FREESECT, (unsigned long)save_first);
+			if (ret < 0) {
+				fdbg("Unable to free sector %d, error = %d\n", save_first, ret);
+				return ret;
+			}
+			save_first = nextsector;
+		}
+	}
 
-int smartfs_extendfile(FAR struct smartfs_mountpt_s *fs, FAR struct smartfs_ofile_s *sf, off_t length)
-{
-	/* TODO This will be added later */
-	return OK;
+	if (result == OK) {
```

Comment:
in my opinion, could you change if to else?

Reply:
or 
as below,
```
sf->filepos = savefilepos;
if (result != OK)
{ }
else 
{ }

Reply:
It is good to clearly separate the code between success and failure for readability.
could you check that again?

Reply:
I'll take a look. Pending.